apps/src/aichat/views/modelCustomization/PublishNotes.tsx

```diff
@@ -70,18 +81,60 @@ const PublishNotes: React.FunctionComponent = () => {
             );
           })}
         </div>
-      )}
-      <div className={styles.footerButtonContainer}>
+      </div>
+      <div className={modelCustomizationStyles.footerButtonContainer}>
+        <Button
+          text="Save"
+          iconLeft={{iconName: 'download'}}
+          type="secondary"
+          color="black"
+          disabled={isDisabled(visibility)}
+          onClick={onSave}
+          className={modelCustomizationStyles.updateButton}
+        />
         <Button
           text="Publish"
           iconLeft={{iconName: 'upload'}}
-          disabled={isDisabled(visibility)}
-          onClick={onUpdate}
-          className={styles.updateButton}
+          disabled={isDisabled(visibility) || !hasFilledOutModelCard}
+          onClick={onPublish}
+          className={modelCustomizationStyles.updateButton}
         />
       </div>
     </div>
   );
 };
 
+const getInputTag = (property: keyof ModelCardInfo) => {
+  return property === 'botName' ? 'input' : 'textarea';
+};
+
+const renderPublishOkNotification = () => {
```

Comment:
I don't exactly know what the performance implications are, but thoughts on just making these components instead of functions that create components every time the parent component is rendered?

Reply:
hmm good q, I think I just had these returning HTML originally before refactoring :) I think I moved them out of the render function just for readability (and I just added some more logic to cover an edge case [Alice identified here](https://github.com/code-dot-org/code-dot-org/pull/57949#discussion_r1564107940)). Is there a difference between this and an inline ternary that returns a component on each render? Isn't the component created each time in that case as well, or is React smarter in that case and keeps them around or something?

Reply:
Yeah, I think the key difference is using functions to render vs values? I don't think it matters a whole lot since these are small, but because they're static components, I was thinking we could benefit from memoizing them. I think we can either just make them static values if they're outside the main component (`const PublishOkNotification = <PublishNotification ... props />`), or use `useMemo` inside the component (`const publishOkNotification = useMemo(renderPublishOkNotification))`)

Reply:
Mentioned at https://github.com/code-dot-org/code-dot-org/pull/57949#discussion_r1564100432, but if a separate component, could use in other customization panels like 'Retrieval' (if retrieval context added/deleted but not updated yet).

Reply:
I this is already ready as a reusable component, just would need to be renamed from `PublishStatus`.

d'oh re: static value / no args, updated

Reply:
Awesome - thanks!